pkg/clusteragent/admission/mutate/auto_instrumentation.go

```diff
@@ -110,9 +112,78 @@ var (
+// The filter excludes two namespaces by default: "kube-system" and the
+// namespace where datadog is installed.
+// Cases:
+// - No enabled namespaces and no disabled namespaces: inject in all namespaces
+// except the 2 namespaces excluded by default.
+// - Enabled namespaces and no disabled namespaces: inject only in the
+// namespaces specified in the list of enabled namespaces. If one of the
+// namespaces excluded by default is included in the list, it will be injected.
+// - Disabled namespaces and no enabled namespaces: inject only in the
+// namespaces that are not included in the list of disabled namespaces and that
+// are not one of the ones disabled by default.
+// - Enabled and disabled namespaces: return error.
+func apmSSINamespaceFilter() (*containers.Filter, error) {
+	apmEnabledNamespaces := config.Datadog.GetStringSlice("apm_config.instrumentation.enabled_namespaces")
+	apmDisabledNamespaces := config.Datadog.GetStringSlice("apm_config.instrumentation.disabled_namespaces")
+
+	if len(apmEnabledNamespaces) > 0 && len(apmDisabledNamespaces) > 0 {
+		return nil, fmt.Errorf("apm.instrumentation.enabled_namespaces and apm.instrumentation.disabled_namespaces configuration cannot be set together")
+	}
+
+	// Prefix the namespaces as needed by the containers.Filter.
+	prefix := containers.KubeNamespaceFilterPrefix
+	apmEnabledNamespacesWithPrefix := make([]string, len(apmEnabledNamespaces))
+	apmDisabledNamespacesWithPrefix := make([]string, len(apmDisabledNamespaces))
+
+	for i := range apmEnabledNamespaces {
+		apmEnabledNamespacesWithPrefix[i] = prefix + apmEnabledNamespaces[i]
+	}
+	for i := range apmDisabledNamespaces {
+		apmDisabledNamespacesWithPrefix[i] = prefix + apmDisabledNamespaces[i]
+	}
+
+	disabledByDefault := []string{
+		prefix + "kube-system",
+		prefix + apiServerCommon.GetResourcesNamespace(),
+	}
+
+	var filterExcludeList []string
+	if len(apmEnabledNamespacesWithPrefix) > 0 && len(apmDisabledNamespacesWithPrefix) == 0 {
+		// In this case, we want to include only the namespaces in the enabled list.
+		// In the containers.Filter, the include list is checked before the
+		// exclude list, that's why we set the exclude list to all namespaces.
+		filterExcludeList = []string{prefix + ".*"}
+	} else {
+		filterExcludeList = append(apmDisabledNamespacesWithPrefix, disabledByDefault...)
+	}
+
+	return containers.NewFilter(containers.GlobalFilter, apmEnabledNamespacesWithPrefix, filterExcludeList)
+}
+
 // InjectAutoInstrumentation injects APM libraries into pods
-func InjectAutoInstrumentation(rawPod []byte, _ string, ns string, _ *authenticationv1.UserInfo, dc dynamic.Interface, _ k8s.Interface) ([]byte, error) {
-	return mutate(rawPod, ns, injectAutoInstrumentation, dc)
+func (apmInstrumentation *APMInstrumentation) InjectAutoInstrumentation(rawPod []byte, _ string, ns string, _ *authenticationv1.UserInfo, dc dynamic.Interface, _ k8s.Interface) ([]byte, error) {
```

Comment:
if we name it `APMInstrumentationWebhook` maybe we can shorter the instance name


```suggestion
func (w *APMInstrumentationWebhook) InjectAutoInstrumentation(rawPod []byte, _ string, ns string, _ *authenticationv1.UserInfo, dc dynamic.Interface, _ k8s.Interface) ([]byte, error) {
```

Reply:
Done :+1: